packages/fuel-indexer-macros/src/schema.rs

```diff
@@ -350,22 +361,75 @@ fn process_type_def(
                     #to_row
                     #decoder
                 };
+
+                let unwrap_or_def = if field_type.nullable {
+                    if EXTERNAL_FIELD_TYPES.contains(field_typ_name.as_str()) {
+                        unwrap_or_default_for_external_type(field_typ_name.clone())
+                    } else {
+                        quote! { .unwrap_or_default() }
+                    }
+                } else {
+                    quote! {}
+                };
+
+                let to_bytes = if EXTERNAL_FIELD_TYPES.contains(field_typ_name.as_str()) {
+                    to_bytes_method_for_external_type(field_typ_name.clone())
+                } else if !ASREF_BYTE_TYPES.contains(field_typ_name.as_str()) {
+                    quote! { .to_le_bytes() }
+                } else {
+                    quote! {}
+                };
+
+                if field_set.contains(&"id".to_string())
+                    && !INTERNAL_INDEXER_ENTITIES.contains(object_name.as_str())
+                    && field_name != *"id"
+                {
+                    parameters = quote! { #parameters #field_name: #typ_tokens, };
+
+                    if !NONDIGESTIBLE_FIELD_TYPES.contains(field_typ_name.as_str()) {
+                        hasher = quote! { #hasher.chain_update(#field_name #clone #unwrap_or_def #to_bytes)};
+                    }
+
+                    field_construction_for_new_impl = quote! {
+                        #field_construction_for_new_impl
+                        #field_name,
+                    };
+                }
             }
 
             schema
                 .object_field_mappings
                 .insert(object_name.clone(), fields_map);
             let strct = format_ident! {"{}", object_name};
 
-            Some(generate_object_trait_impls(
-                strct,
+            let object_trait_impls = generate_object_trait_impls(
+                strct.clone(),
                 strct_fields,
                 type_id,
                 field_extractors,
                 from_row,
                 to_row,
                 schema.is_native,
-            ))
+            );
+
+            let new_method_impl = if field_set.contains(&"id".to_string()) {
```

Comment:
Can we use `fuel_indexer_database_types::IdCol::to_lowercase_str()` (or something) here?